### PR TITLE
Diff 2.0: impl summary

### PR DESCRIPTION
**Pull request by ZipFast**
### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: sync_diff_inspector/source/basic.go [hunk 1]
```diff
@@ -99,9 +98,19 @@ func getOriginTable(sourceTableMap map[string]*common.TableSource, table *common
 	return originSchema, originTable
 }
 
+func (s *BasicSource) GetSourceTableMap() map[string]*common.TableSource {
```
Comment (3pointer) [inline sync_diff_inspector/source/basic.go]: why add this function?

File: sync_diff_inspector/utils/utils.go [hunk 2]
```diff
@@ -58,6 +59,45 @@ func NewWorkerPool(limit uint, name string) *WorkerPool {
 	}
 }
 
+// Returns either an ascii code, or (if input is an arrow) a Javascript key code.
+func GetChar() (ascii int, keyCode int, err error) {
```
Comment (3pointer) [inline sync_diff_inspector/utils/utils.go:64]: maybe we should ask PM to add this interactive action🤔

File: sync_diff_inspector/diff.go [hunk 3]
```diff
@@ -152,23 +205,77 @@ func (df *Diff) Equal(ctx context.Context) error {
...
+		log.Info("generate chunk", zap.Int("chunk id", c.ChunkRange.ID), zap.String("table", dbutil.TableName(tableDiff.Schema, tableDiff.Table)))
 		pool.Apply(func() {
-			res, err := df.consume(ctx, c)
-			if err != nil {
-				// TODO: catch error
-			}
-			// TODO: handle res
-			if res {
-
+			tableDiff := df.workSource.GetTables()[c.TableIndex]
```
Comment (3pointer) [inline sync_diff_inspector/diff.go]: this function needs a lock.
I suggest save the `schema` `table` into Chunk

File: sync_diff_inspector/source/basic.go (hunk 1 above)
Comment (3pointer) [reply sync_diff_inspector/source/basic.go]: We can implement the `GetTableStruct` in each Source. `SourceTableMap` is too implied

Comment (ZipFast) [reply sync_diff_inspector/source/basic.go]: Fixed

File: sync_diff_inspector/diff.go (hunk 3 above)
Comment (ZipFast) [reply sync_diff_inspector/diff.go]: I cannot see there is any concurrent writes here

File: sync_diff_inspector/diff.go [hunk 4]
```diff
@@ -152,23 +205,64 @@ func (df *Diff) Equal(ctx context.Context) error {
...
+				df.report.SetTableMeetError(schema, table, err)
 			}
 		})
 	}
 
 	return nil
 }
 
+func (df *Diff) compareStruct(ctx context.Context, c *splitter.RangeInfo) (structEq bool, err error) {
+	switch upstream := df.upstream.(type) {
```
Comment (3pointer) [inline sync_diff_inspector/diff.go]: This code can be refine simplify
```
sourceTableInfos, err := upstream.GetSourceStructInfo(ctx, c.TableIndex)
if err != nil {
    return false, errors.Trace(err)
}
structEq = true
for _, tableInfo := range sourceTableInfos {
    eq, _ := dbutil.EqualTableInfo(tableInfo, df.downstream.GetTables()[c.TableIndex].Info)
    structEq = structEq && eq
}
```

File: sync_diff_inspector/diff.go [hunk 5]
```diff
@@ -199,7 +293,10 @@ func (df *Diff) generateChunksIterator(ctx context.Context) (source.RangeIterato
 					zap.Int("id", node.GetID()),
 					zap.Reflect("chunk", node),
 					zap.String("state", node.GetState()))
-				df.cp.SetCurrentSavedID(node.GetID() + 1)
+				df.cp.SetCurrentSavedID(node.GetID())
+			}
+			if err = df.loadReport(path + "_report"); err != nil {
```
Comment (3pointer) [inline sync_diff_inspector/diff.go:289]: 🤔 we'd better put the report at the same level of checkpoint directory. to compare with checkpoint. report is an Independent part.

File: sync_diff_inspector/diff.go [hunk 6]
```diff
@@ -152,23 +205,64 @@ func (df *Diff) Equal(ctx context.Context) error {
...
-			if res {
-
+			tableDiff := df.workSource.GetTables()[c.TableIndex]
+			schema, table := tableDiff.Schema, tableDiff.Table
+			structEq, err := df.compareStruct(ctx, c)
+			if err == nil {
+				if !structEq {
+					df.report.SetTableStructCheckResult(schema, table, structEq)
+				}
+				res, rowsAdd, rowsDelete, rowsCnt, err := df.consume(ctx, c)
```
Comment (3pointer) [inline sync_diff_inspector/diff.go]: config `ignore-data-check` and `ignore-struct-check` can be add here.

File: sync_diff_inspector/diff.go (hunk 3 above)
Comment (3pointer) [reply sync_diff_inspector/diff.go]: ok, There is no race.

but In this code we will repeated compare the same table `chunksCount` times. we need reduce the compare times. 🤔

File: sync_diff_inspector/diff.go (hunk 5 above)
Comment (ZipFast) [reply sync_diff_inspector/diff.go:289]: this report should be a temp file, and will be deleted in the end of diff process

Comment (3pointer) [reply sync_diff_inspector/diff.go:289]: yes, but it's strange to see a file called `checkpoint.pb_report` that has nothing to do with checkpoint.e

**Pull request merged**